PUBDEV-7269 Multinomial AUC/AUCPR

maurever opened the pull request:
# Multinomial AUC/PR AUC

Todo:
- [x]  implement OVR and OVO calculation 
- [x] implement macro and weighted average
- [x] return all metrics in table
- [x] Python API
- [x] R API
- [x] Java test (it returns correct values, grid search test, early stopping test)
- [x] Python, R tests
- [x] add parameter to set default `auc()` and `prauc()` value (for grid search and early stopping)
- [x] set AUC/PR AUC in grid search for multinomial metric
- [x] set AUC/PR AUC in early stopping for multinomial metric
- [x] solve metric construction in GenericModel

- [x] demo
- [x] test performance
- [x] update documentation AUC/PR AUC page
- [x] decide the API is designed correctly 
- [x] decide what is the limit of domain size
- [x] disable by default

## Current API

### 1. User can call `auc()` and `aucpr()` from Python and R with MultinomialMetric object. This method can return aggregated value - default is Weighted OVR.

### 2. User can get whole multinomial AUC/PR AUC summary using `multinomial_auc_table()` and  `multinomial_aucpr_table()` method.

![api](https://user-images.githubusercontent.com/11465784/97172263-154c0d80-178f-11eb-8fa5-2390550bcd26.png)

### 3. Users can change the default method to calculate AUC/PR AUC using `auc_type` parameter in the model. The possibilities are  `NONE`, `MACRO_OVR`, `WEIGHTED_OVR`, `MACRO_OVO`, `WEIGHTED_OVO`,  `AUTO`.

![auc_type](https://user-images.githubusercontent.com/11465784/97172294-209f3900-178f-11eb-835e-059a9f95683c.png)

### 4. The limit of domain to print this summary in H2O logs is 20 (same as for confusion matrix). The limit to calculate multinomial AUC/PR AUC is currently 50.

![too_big](https://user-images.githubusercontent.com/11465784/97172506-64923e00-178f-11eb-9bcc-d3023343ba21.png)

### 5. Calculation of multinomial AUC/PR AUC could be time consuming - it is disabled by default

### 6. User can use `AUC` and `AUCPR` metric for early stopping in multinomial problems.

![early_stopping](https://user-images.githubusercontent.com/11465784/97172912-04e86280-1790-11eb-9c53-54b593d38f40.png)

### 7. User can get `auc` and `aucpr` metric in grid search and can sort result models using this metric.

![gridsearch](https://user-images.githubusercontent.com/11465784/97173499-f6e71180-1790-11eb-9563-a9015669cbf0.png)

## Performance

100 trees **without** calculating AUC/PR AUC: **0 min 30.805 sec (~ 31 sec)**
100 trees **with** calculating      AUC/PR AUC: **4 min 20.068 sec (~264 sec)**

![performance](https://user-images.githubusercontent.com/11465784/97185909-f35b8680-17a0-11eb-96dc-df7722ddc00c.png)

JIRA: https://0xdata.atlassian.net/jira/software/c/projects/PUBDEV/issues/PUBDEV-7269

neemah2o reviewed (approved):
LGTM

michalkurka commented on h2o-core/src/test/java/water/rapids/ast/prims/advmath/AstUniqueTest.java:
```diff
@@ -124,5 +125,30 @@ public void categoricalNasRecognized() {
   }
 
+  @Test
+  public void stringsUnsupported() {
```
This looks like a wrong merge. This code is already in master.

michalkurka commented:
Please make sure you rebase this branch - currently it looks like you are changing files that this PR shouldn't be changing - eg. TransformWrappedVec.java

michalkurka commented:
Otherwise it IMHO looks good -thank you for this contribution!

maurever commented:
> Please make sure you rebase this branch - currently it looks like you are changing files that this PR shouldn't be changing - eg. TransformWrappedVec.java

Ou no, you are true! I rebased on rel-zermelo but this PR is on master. Fixed.

michalkurka reviewed (approved):
LGTM!!!

maurever commented:
Fix AutoML leaderboard and tests, the failed tests look unrelated and are on the master branch too. So I am merging.
